Task/Des-1792 add public viewer

nathanfranklin opened the pull request:
## Overview: ##

Add public viewer (requires https://github.com/TACC-Cloud/geoapi/pull/53)

This also refactors part of the authentication process:
* redirects to calling url after successful login
* fixes bug where we were logged occurred twice
* adds /logout route to log out users (for testing 👍 )

## PR Status: ##

* [x] Ready.

## Related Jira tickets: ##

* [DES-1792](https://jira.tacc.utexas.edu/browse/DES-1792)
* [DES-1791](https://jira.tacc.utexas.edu/browse/DES-1791)

## Summary of Changes: ##
* adds public route
* remove/modify components when viewing public map to hide/disable users editing map
* add JWT to request only if user is logged in (for local dev only)
* redirects to calling url after successful login
* fixes bug where we were logged occurred twice
* adds /logout route to log out users
* add "login" button


## Testing Steps: ##
**Check access to private map**
1. Create a map
2. Make the map public (see UI Photos below)
3. Go to the public url link while logged in
3. Go to the public url link while not logged in
3. Make private and confirm that you can't access as public

**Check that user cannot modify map **
1. Make the map public (see UI Photos below)
2. Go to the public url link while logged in or out, and confirm that you can't modify the map in anyway.

**Check redirect after login**
1. logout  (`http://localhost:4200/logout`)
2. Go directly to private route like `http://localhost:4200/project/4577fb69-15ff-41ab-a8fd-076e83500daa`
3. Ensure that after logging in, you end up at http://localhost:4200/project/4577fb69-15ff-41ab-a8fd-076e83500daa


## UI Photos:

![Screen Shot 2021-04-05 at 3 25 06 PM](https://user-images.githubusercontent.com/8287580/113624589-72dd6380-9625-11eb-9ef3-53dc229a25c8.png)
![Screen Shot 2021-04-05 at 3 25 21 PM](https://user-images.githubusercontent.com/8287580/113624588-72dd6380-9625-11eb-96ea-56485e64f4bc.png)
![Screen Shot 2021-04-05 at 3 25 30 PM](https://user-images.githubusercontent.com/8287580/113624587-72dd6380-9625-11eb-9d5e-effae3c0b7c0.png)
![Screen Shot 2021-04-05 at 3 25 41 PM](https://user-images.githubusercontent.com/8287580/113624586-7244cd00-9625-11eb-9f4c-8ee27f94cab5.png)
![Screen Shot 2021-04-05 at 3 25 48 PM](https://user-images.githubusercontent.com/8287580/113624583-7244cd00-9625-11eb-8f13-03fc50366f7b.png)
![Screen Shot 2021-04-05 at 3 26 07 PM](https://user-images.githubusercontent.com/8287580/113624582-71ac3680-9625-11eb-95be-72096a41a573.png)
![Screen Shot 2021-04-05 at 3 26 26 PM](https://user-images.githubusercontent.com/8287580/113624579-71ac3680-9625-11eb-8dbf-7df27983014b.png)


## Todo: ##
- [x] do not subscribe to notifications if anonymous
- [x] do not get project users if anonymous
- [x] remove login for public route
- [x] show top-level notice in public view for public map that is missing or is private.
- [x] clarification: ~do we need login from public view (like a "go here to login and create/edit your own maps")~ Joe: Hazmapper logo could link to private area
- [x] clarification: is dialog confirming public/private change needed?


Created tickets
- Added ticket to fix panels styling (users scroll, spacing..) ( https://jira.tacc.utexas.edu/browse/DES-1923)
- Added ticket add follow on ticket for public users to mess with tile layers but not to update them (https://jira.tacc.utexas.edu/browse/DES-1924)
- Added ticket for link to private view from public view if logged in and user has access to that project (https://jira.tacc.utexas.edu/browse/DES-1921)
- Added ticket for Ian's suggested `Duplicate this map` button and way to request access to private version of app). ( https://jira.tacc.utexas.edu/browse/DES-1922)

duckonomy reviewed (approved):
All good. But there is one case where you 1. Create a map public. 1.5 Copy the public map link 2. Logout 3. Go to that public map. And you get a 403 for each asset calls to the back end with a blank map and the loader spinner.

duckonomy commented on src/app/components/main/main.component.ts:
```diff
@@ -1,31 +1,34 @@
+    if (this.authService.isLoggedIn()) {
+      this.authService.getUserInfo();
+    }
     this.authService.currentUser.subscribe(next => this.currentUser = next);
```
`this.authService.currentUser.subscribe(next => this.currentUser = next ? next : undefined);`
This gets rid of the exception on logout

nathanfranklin commented:
> All good. But there is one case where you 1. Create a map public. 1.5 Copy the public map link 2. Logout 3. Go to that public map. And you get a 403 for each asset calls to the back end with a blank map and the loader spinner.

fixed this by fixing an error in TACC-Cloud/geoapi#53 (see commit https://github.com/TACC-Cloud/geoapi/pull/53/commits/20e54ac98b33a8e39104d06863f10da120efa664)

Also added https://jira.tacc.utexas.edu/browse/DES-1927 as there is one more scenario that needs to be fixed.
